// import-sort-config/lib/index.js
"use strict";
var assign = require("core-js/library/fn/object/assign");
var minimatch = require("minimatch");
var fs_1 = require("fs");
var resolve = require("resolve-from");
var findRoot = require("find-root");
;
exports.DEFAULT_CONFIGS = {
    ".js, .jsx, .es6, .es": {
        parser: "babylon",
        style: "eslint",
    },
    ".ts, .tsx": {
        parser: "typescript",
        style: "eslint",
    },
};
function getConfig(extension, directory) {
    var defaultConfig = getConfigForExtension(exports.DEFAULT_CONFIGS, extension);
    var packageConfig;
    if (directory) {
        packageConfig = getConfigFromPackage(directory, extension);
    }
    var actualConfig = mergeConfigs([defaultConfig, packageConfig]);
    if (!actualConfig) {
        return;
    }
    return resolveConfig(actualConfig, directory);
}
exports.getConfig = getConfig;
function getConfigFromPackage(directory, extension) {
    var packageConfigs = getConfigsFromPackage(directory);
    if (!packageConfigs) {
        return;
    }
    return getConfigForExtension(packageConfigs, extension);
}
function getConfigForExtension(configs, extension) {
    var foundConfigs = Object.keys(configs).map(function (joinedGlobs) {
        var globs = joinedGlobs.split(",").map(function (rawGlob) { return rawGlob.trim(); });
        var config = configs[joinedGlobs];
        if (globs.some(function (glob) { return minimatch(extension, glob); })) {
            return config;
        }
    });
    return mergeConfigs(foundConfigs);
}
function getConfigsFromPackage(directory) {
    try {
        return JSON.parse(fs_1.readFileSync(findRoot(directory) + "/package.json", "utf-8")).importSort;
    }
    catch (e) {
        return;
    }
}
function mergeConfigs(rawConfigs) {
    var configs = rawConfigs.filter(function (rawConfig) { return !!rawConfig; });
    if (configs.length === 0) {
        return;
    }
    return configs.reduce(function (previousConfig, currentConfig) {
        if (!currentConfig) {
            return previousConfig;
        }
        var config = assign({}, previousConfig);
        if (currentConfig.parser) {
            config.parser = currentConfig.parser;
        }
        if (currentConfig.style) {
            config.style = currentConfig.style;
        }
        return config;
    });
}
function resolveConfig(config, directory) {
    var resolvedConfig = {
        config: config,
    };
    if (config.parser) {
        resolvedConfig.parser = resolveParser(config.parser, directory);
    }
    if (config.style) {
        resolvedConfig.style = resolveStyle(config.style, directory);
    }
    return resolvedConfig;
}
function resolveParser(module, directory) {
    if (module.indexOf(".") === 0) {
        return resolveModule(module, directory);
    }
    if (module.indexOf("import-sort-parser") === 0) {
        return resolveModule(module, directory);
    }
    return resolveModule("import-sort-parser-" + module, directory);
}
function resolveStyle(module, directory) {
    if (module.indexOf(".") === 0) {
        return resolveModule(module, directory);
    }
    if (module.indexOf("import-sort-style-") === 0) {
        return resolveModule(module, directory);
    }
    return resolveModule("import-sort-style-" + module, directory);
}
function resolveModule(module, directory) {
    if (directory) {
        var path_1 = resolve(directory, module);
        if (path_1) {
            return path_1;
        }
    }
    var path = resolve(__dirname, module);
    if (path) {
        return path;
    }
}
